File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
import sys
read = sys.stdin.readline

def dfs(num):
    if len(arr) == M:
        print(' '.join(map(str, arr)))
        return

    for i in range(num + 1, N+1):
        if i not in arr:
            arr.append(i)
            dfs(i)
            arr.pop()

N, M = map(int, read().split())
arr = []

dfs(0)

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.util.*;
import java.io.*;

class Main {
    static int N, M;
    static int[] arr;
    static StringBuilder sb = new StringBuilder();
    
    public static void dfs(int num, int idx) {
        if(idx == M) {
            for(int n: arr) sb.append(n + " ");
            sb.append("\n");
            return;
        }
        
        for(int i=num;i<=N;i++) {
            arr[idx] = i;
            dfs(i+1, idx+1);
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[M];
        
        dfs(1, 0);
        
        System.out.print(sb);
    }
}